File: gt4py/linear_advection/compute_mass.py
import numpy as np

# Function to compute mass matrix and its inverse
def compute_mass(phi,wts2d,nx,ny,r,hx,hy,y_c,pts2d_y,eq_type) :
    n_qp = len(pts2d_y)
    # determ  = hx*hy/4
    # mass = np.zeros((nx, ny))
    # cos_factor = np.zeros((nx, ny))
    # sin_factor = np.zeros((nx, ny))
    # inv_mass = np.zeros((nx, ny))

    cos_factor = np.ones((nx, ny, 1, n_qp))
    sin_factor = np.zeros((nx, ny, 1, n_qp))

    for j in range(ny):
        if (eq_type == "linear" or eq_type == "swe") :
            # cos_factor = np.ones((n_qp, n_qp))
            pass
        elif (eq_type == "adv_sphere" or eq_type == "swe_sphere") :
            cos_y = np.cos( y_c[j]+pts2d_y*hy/2 ) 
            sin_y = np.sin( y_c[j]+pts2d_y*hy/2 ) 
            cos_factor[:,j,0,:] = np.tile(cos_y, (nx, 1))
            sin_factor[:,j,0,:] = np.tile(sin_y, (nx, 1))
        else :
            print (eq_type,"unsupported option, using linear ")
            cos_factor[j] = np.ones((n_qp, n_qp))

    dim = (r+1)**2
    determ  = hx*hy/4
    mass = np.zeros((nx, ny, 1, dim, dim))
    inv_mass = np.zeros((nx, ny, 1, dim, dim))
    for j in range(ny):
        for i in range(nx):
            # r_loc=rdist[i,j]
            # dim = int((r_loc+1)*(r_loc+1))
            matrix = np.zeros((dim,dim))
            for m in range(dim):
                for n in range(dim):
                    # det*sum(i-th basis function in qp * j-th basis function in qp * metric factor * weights)
                    # matrix[m,n]=determ * np.dot(wts2d, (phi[r_loc][:,m] * phi[r_loc][:,n] * cos_factor[j]))
                    matrix[m,n]=determ * np.dot(wts2d, (phi[:,m] * phi[:,n] * cos_factor[i,j,0]))
            mass[i,j,0,:,:] = matrix
            inv_mass[i,j,0,:,:] = np.linalg.inv(mass[i,j])

    return mass, inv_mass

File: matlab/compute_mass.m
function [mass,inv_mass] =compute_mass(phi,wts2d,d1,d2,r,hx,hy,factor)
%compute mass matrix and its inverse, different in all elements due to cosine factors
%M(i,j)=integral_K(Phi_j*Phi_i*dA)

dim=(max(r(:))+1)^2;

% TODO: shouldn't this be zeros?
%dimensions: (cardinality)x(cardinality)x(num_elems)
mass=repmat(eye(dim,dim),1,1,d1*d2);
inv_mass=repmat(eye(dim,dim),1,1,d1*d2);


% mass=repmat(zeros(dim,dim),1,1,d1*d2);
% inv_mass=repmat(zeros(dim,dim),1,1,d1*d2);



%determinant of the affine mapping from reference to physical element (this
%is assumed to be constant)
determ=hx*hy/4;

for k=1:d1*d2
    
    elem_x=floor((k-1)/d2)+1; % element x index
    elem_y=mod((k-1),d2)+1; % element y index
    r_loc=r(elem_y,elem_x);
    
    for i=1:(r_loc+1)^2
        for j=1:(r_loc+1)^2
            %det*sum(i-th basis function in qp * j-th basis function in qp * metric
            %factor * weights)
            mass(i,j,k)=determ*wts2d'*(phi{r_loc}(:,i).*phi{r_loc}(:,j).*factor(:,k));
%             mass(i,j,k)=determ*wts2d'*(phi{r_loc}(:,i).*phi{r_loc}(:,j));

        end
    end
    
    inv_mass(1:(r_loc+1)^2,1:(r_loc+1)^2,k)=mass(1:(r_loc+1)^2,1:(r_loc+1)^2,k)\eye((r_loc+1).^2);
    
end

end